Nama : M. Hafiz Attariq<br>
NIM : 24060121140174<br>
Nama Dataset : Nursery<br>
Link Dataset : https://archive.ics.uci.edu/ml/machine-learning-databases/nursery/nursery.data

# Import Libraries

In [45]:
# Import Libraries
import pandas as pd
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Memuat Dataset

In [46]:
# Memuat Dataset
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/nursery/nursery.data"

names = ['parents', 'has_nurs', 'form', 'children', 'housing', 'finance', 'social', 'health', 'class']

dataset = pd.read_csv(url, names=names)

## Preprocess the data

In [47]:
le = LabelEncoder()
for col in dataset.columns:
    dataset[col] = le.fit_transform(dataset[col])

# Dimensi Dataset

In [48]:
dataset.shape

(12960, 9)

jumlah baris dalam dataset adalah 12960 dan jumlah kolom dalam data set adalah 9

# Isi Dataset

In [49]:
dataset

parents  has_nurs  form  children  housing  finance  social  health  \
0            2         3     0         0        0        0       0       2   
1            2         3     0         0        0        0       0       1   
2            2         3     0         0        0        0       0       0   
3            2         3     0         0        0        0       2       2   
4            2         3     0         0        0        0       2       1   
...        ...       ...   ...       ...      ...      ...     ...     ...   
12955        0         4     2         3        1        1       2       1   
12956        0         4     2         3        1        1       2       0   
12957        0         4     2         3        1        1       1       2   
12958        0         4     2         3        1        1       1       1   
12959        0         4     2         3        1        1       1       0   

       class  
0          2  
1          1  
2          0  
3          2  
4          1  
...      ...  
12955      3  
12956      0  
12957      3  
12958      3  
12959      0  

[12960 rows x 9 columns]

Diatas merupakan tabel dari isi dataset

# Distribusi Kelas Data


In [50]:
dataset.groupby('class').size()

class
0    4320
1    4266
2       2
3    4044
4     328
dtype: int64

### Pertemuan 2

## Import Model Selection

In [51]:
from sklearn import model_selection

##Train-Test Split

In [52]:
array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

X berisi fitur (features) dari dataset, yaitu kolom 1 hingga 4.<br>
Y berisi target (label) dari dataset, yaitu kolom 5.<br><br>
train_test_split adalah fungsi dari modul model_selection yang digunakan untuk membagi dataset menjadi data latih dan data uji.<br>
X_train dan Y_train adalah data latih untuk fitur dan target.<br>
X_validation dan Y_validation adalah data uji untuk fitur dan target.<br>
test_size=0.20 menunjukkan bahwa 20% dari dataset akan digunakan sebagai data uji.<br>
random_state digunakan untuk mengatur seed pada proses pembagian data agar hasilnya dapat direproduksi.

##  Menampilkan shape dari masing-masing subset data setelah dilakukan pembagian.

In [53]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (10368, 4)
X_validation : (2592, 4)
Y_train : (10368,)
Y_validation : (2592,)


## K-Folds Cross Validation

In [54]:
seed = 7

scoring = 'accuracy'

## Model Algoritma di Modul

In [55]:
#Import library algoritma
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB

#Spot Check Algorithms
models = []

models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

## Evaluasi Algoritma Model

In [56]:
# evaluate each model in turn
results = []

names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train,
  cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.307965 (0.010257)
NB: 0.321761 (0.011645)
SVM: 0.287422 (0.010754)


Mengevaluasi kinerja beberapa model machine learning menggunakan validasi silang (cross-validation).<br>
Model yang dinilai:<br>


*   K-Nearest Neighbors (KNN): Akurasi rata-rata sekitar 30.79%, dengan deviasi standar sekitar 0.010257.
*   Naive Bayes (NB): Akurasi rata-rata sekitar 32.17%, dengan deviasi standar sekitar 0.011645.
*   Support Vector Machine (SVM): Akurasi rata-rata sekitar 28.74%, dengan deviasi standar sekitar 0.010754.

## Import Library untuk Testing

In [57]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## Melakukan pengujian tentang keakuratan model Naive Bayes terhadap data

In [58]:
# Make predictions on validation dataset
NB = GaussianNB()

NB.fit(X_train, Y_train)

predictions = NB.predict(X_validation)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.30671296296296297
[[188 462 218]
 [211 401 232]
 [213 461 206]]
              precision    recall  f1-score   support

           0       0.31      0.22      0.25       868
           1       0.30      0.48      0.37       844
           2       0.31      0.23      0.27       880

    accuracy                           0.31      2592
   macro avg       0.31      0.31      0.30      2592
weighted avg       0.31      0.31      0.30      2592



karena Naive Bayes memiliki nilai akurasi perkiraan terbesar, maka dilakukan pengujian mengenai keakuratan Model tersebut.<br>
Akurasi model pada dataset validasi adalah sekitar 30.67%.